fontforge/autosave.c

```diff
@@ -25,136 +25,121 @@
+static char *GetAutoDirName() {
+    char *userdir = getFontForgeUserDir(Config);
+    char *autodir = NULL;
+    
+    if (userdir != NULL) {
+        int ret = asprintf(&autodir, "%s/autosave", userdir);
+        free(userdir);
+        if (ret == -1) {
+            return NULL;
+        } else if (!GFileExists(autodir) && GFileMkDir(autodir) == -1) {
+            free(autodir);
+            return NULL;
+        }
     }
-    return( dir );
+    return autodir;
 }
 
-static void MakeAutoSaveName(SplineFont *sf) {
+static bool MakeAutoSaveName(SplineFont *sf) {
     char buffer[1025];
     char *autosavedir;
     static int cnt=0;
 
-    if ( sf->autosavename )
-return;
-    autosavedir = getAutoDirName(buffer);
-    if ( autosavedir==NULL )
-return;
-    while ( 1 ) {
-	sprintf( buffer, "%s/auto%06x-%d.asfd", autosavedir, getpid(), ++cnt );
-	if ( access(buffer,F_OK)==-1 ) {
-	    sf->autosavename = copy(buffer);
-return;
-	}
+    if (!sf->autosavename && (autosavedir = GetAutoDirName())) {
+        while (true) {
+            int nret = snprintf(buffer, sizeof(buffer), "%s/auto%06x-%d.asfd",
+                                autosavedir, getpid(), ++cnt);
+            if (nret < 0 || nret >= sizeof(buffer)) {
+                break;
+            } else if (!GFileExists(buffer)) {
+                sf->autosavename = copy(buffer);
+                break;
+            }
+        }
+        free(autosavedir);
+    }
+    return (sf->autosavename != NULL);
+}
+
+static void _DoAutoSaves(FontViewBase *fvs) {
+    FontViewBase *fv;
+    SplineFont *sf;
+
+    if (AutoSaveFrequency <= 0)
+        return;
+
+    for (fv = fvs; fv != NULL; fv = fv->next ) {
+        sf = (fv->cidmaster ? fv->cidmaster : fv->sf);
+        if (sf->changed_since_autosave && MakeAutoSaveName(sf)) {
+            SFAutoSave(sf, fv->map);
+        }
     }
 }
 
+void DoAutoSaves(void) {
+    _DoAutoSaves(FontViewFirst());
+}
 
-int DoAutoRecoveryExtended(int inquire)
-{
+int DoAutoRecoveryExtended(int inquire) {
     char buffer[1025];
-    char *recoverdir = getAutoDirName(buffer);
-    DIR *dir;
-    struct dirent *entry;
+    char *recoverdir = GetAutoDirName();
+    GDir *dir;
+    const gchar *entry;
     int any = false;
     SplineFont *sf;
     int inquire_state=0;
 
-    if ( recoverdir==NULL )
-return( false );
-    if ( (dir = opendir(recoverdir))==NULL )
-return( false );
-    while ( (entry=readdir(dir))!=NULL ) {
-	if ( strcmp(entry->d_name,".")==0 || strcmp(entry->d_name,"..")==0 )
-    continue;
-	sprintf(buffer,"%s/%s",recoverdir,entry->d_name);
-	fprintf( stderr, "Recovering from %s... ", buffer);
-	if ( (sf = SFRecoverFile(buffer,inquire,&inquire_state)) ) {
-	    any=true;
-	    if ( sf->fv==NULL )		/* Doesn't work, cli arguments not parsed yet */
-		FontViewCreate(sf,false);
-	    fprintf( stderr, " Done\n" );
-	}
+    if (recoverdir == NULL) {
+        return false;
+    } else if ((dir = g_dir_open(recoverdir, 0, NULL)) == NULL) {
```

Comment:
On Windows, opendir won't work if the path contains nonstandard characters (characters outside of their current codepage). GLib ensures that the paths are UTF-8 encoded (on Windows) so this will work irregardless of what current codepage the user has.